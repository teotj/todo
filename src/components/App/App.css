@mixin center {
  display: block;
  margin-left: auto;
  margin-right: auto;
}

.root {
  padding: 20px 28px;
}

.header {
  text-align: center;
  margin-bottom: 20px;
}

.formGroup {
  width: 500px;
  margin: auto;
}

.todoItem {
}

.add {
  @include center;

  margin-top: 10px;
}

.clear {
}

.todoTable {
  @include center;

  width: 70%;
  margin-top: 20px;
  background-color: #f5f8fa;
  padding: 5px 20px 20px;
  border-radius: 6px;
}

.tabs {
}

.tab[role='tabList'] {
  cursor: pointer;
}
