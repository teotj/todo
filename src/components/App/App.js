import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Button, FormGroup, FocusStyleManager, H3, InputGroup, Intent, Tabs, Tab, Tooltip } from '@blueprintjs/core';
import { IconNames } from '@blueprintjs/icons';

import TodoTable from 'components/TodoTable';
import ErrorBoundary from 'components/ErrorBoundary';
import { actions } from 'store';

import css from './App.css';

FocusStyleManager.onlyShowFocusOnTabs();

export class App extends PureComponent {
  state = {
    itemValue: ''
  };

  onAddItem = () => {
    const { onAddItem } = this.props;
    const { itemValue } = this.state;

    onAddItem(itemValue);

    this.handleClearItemValue();
  };

  onCompleteItem = (e, id) => {
    const { onCompleteItem } = this.props;

    onCompleteItem(id);
  };

  onDeleteItem = (e, id) => {
    const { onDeleteItem } = this.props;

    onDeleteItem(id);
  };

  handleItemValueChange = (e) => this.setState({ itemValue: e.target.value });

  handleClearItemValue = () => this.setState({ itemValue: '' });

  renderTodoInput = () => {
    const { itemValue } = this.state;

    const isBlank = itemValue === '';
    const clearButton = (
      <Tooltip content="Clear input">
        <Button
          minimal
          icon={IconNames.CROSS}
          className={css.clear}
          onClick={this.handleClearItemValue}
        />
      </Tooltip>
    );
    const addButton = (
      <Button
        text="Add todo"
        disabled={isBlank}
        icon={IconNames.ADD}
        intent={Intent.SUCCESS}
        className={css.add}
        onClick={this.onAddItem}
      />
    );

    return (
      <React.Fragment>
        <ErrorBoundary>
          <FormGroup className={css.formGroup}>
            <InputGroup
              large
              value={itemValue}
              placeholder="Add todo here..."
              onChange={this.handleItemValueChange}
              className={css.todoItem}
              rightElement={isBlank ? null : clearButton}
            />
          </FormGroup>
          {addButton}
        </ErrorBoundary>
      </React.Fragment>
    );
  };

  renderTodoItems = () => {
    const { items } = this.props;

    const completedItems = items.filter(({ isCompleted }) => isCompleted);
    const activeItems = items.filter(({ isCompleted }) => !isCompleted);
    const tabs = [
      {
        id: 'all',
        title: 'All',
        panel: <TodoTable items={items} onComplete={this.onCompleteItem} onDelete={this.onDeleteItem} />,
        className: css.tab
      },
      {
        id: 'completed',
        title: 'Completed',
        panel: <TodoTable items={completedItems} onComplete={this.onCompleteItem} onDelete={this.onDeleteItem} />,
        className: css.tab
      },
      { id: 'active',
        title: 'Active',
        panel: <TodoTable items={activeItems} onComplete={this.onCompleteItem} onDelete={this.onDeleteItem} />,
        className: css.tab
      }
    ];

    return (
      <div className={css.todoTable}>
        <ErrorBoundary>
          <Tabs
            id="todoTabs"
            onChange={this.handleTabChange}
            className={css.tabs}
            renderActiveTabPanelOnly
          >
            {tabs.map((tab) => (
              <Tab key={tab.id} {...tab} />
            ))}
          </Tabs>
        </ErrorBoundary>
      </div>
    );
  };

  render() {
    return (
      <div className={css.root}>
        <H3 className={css.header}>Todo</H3>
        {this.renderTodoInput()}
        {this.renderTodoItems()}
      </div>
    );
  }
}

App.defaultProps = {
  items: []
};

App.propTypes = {
  items: PropTypes.arrayOf(PropTypes.shape({})),
  onAddItem: PropTypes.func.isRequired,
  onCompleteItem: PropTypes.func.isRequired,
  onDeleteItem: PropTypes.func.isRequired
};

export const mapStateToProps = ({ items }) => ({ items });

export const mapDispatchToProps = (dispatch) => ({
  onAddItem: (item) => dispatch(actions.addItem(item)),
  onCompleteItem: (id) => dispatch(actions.completeItem(id)),
  onDeleteItem: (id) => dispatch(actions.deleteItem(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(App);
