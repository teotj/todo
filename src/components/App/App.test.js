import React from 'react';
import { shallow } from 'enzyme';

import { App, mapStateToProps, mapDispatchToProps } from './App';

const renderApp = (props = {}) => {
  const defaultItems = [
    { title: 'TodoA', id: '-idA-', isCompleted: true },
    { title: 'TodoB', id: '-idB-', isCompleted: false }
  ];
  const defaultOnAddItem = () => {};
  const defaultOnCompleteItem = () => {};
  const defaultOnDeleteItem = () => {};

  return (
    <App
      onAddItem={defaultOnAddItem}
      onCompleteItem={defaultOnCompleteItem}
      onDeleteItem={defaultOnDeleteItem}
      items={defaultItems}
      {...props}
    />
  );
};

const shallowRenderApp = (props) => shallow(renderApp(props));

describe('<App>', () => {
  it('Should render', () => {
    const actual = shallowRenderApp();
    expect(actual).toMatchSnapshot();
  });

  it('Should update item value on change', () => {
    const actual = shallowRenderApp();
    actual.find('.todoItem').simulate('change', { target: { value: 'TodoA' } });
    expect(actual.state().itemValue).toBe('TodoA');
  });

  it('Should clear input value on when clear button is clicked', () => {
    const actual = shallowRenderApp();
    actual.setState({ itemValue: 'TodoB' });
    expect(actual.state().itemValue).toBe('TodoB'); // Making sure itemValue is not an empty string
    actual.instance().handleClearItemValue();
    expect(actual.state().itemValue).toBe('');
  });

  it('Should call onAddItem', () => {
    const onAddItem = jest.fn();
    const actual = shallowRenderApp({ onAddItem });
    actual.setState({ itemValue: 'TodoC' });
    actual.instance().onAddItem();
    expect(onAddItem).toBeCalledWith('TodoC');
    expect(actual.state().itemValue).toBe('');
  });

  it('Should call onDeleteItem', () => {
    const onDeleteItem = jest.fn();
    const fakeEvent = {};
    const actual = shallowRenderApp({ onDeleteItem });
    actual.instance().onDeleteItem(fakeEvent, '-idA-');
    expect(onDeleteItem).toBeCalledWith('-idA-');
  });

  it('Should call onCompleteItem', () => {
    const onCompleteItem = jest.fn();
    const fakeEvent = {};
    const actual = shallowRenderApp({ onCompleteItem });
    actual.instance().onCompleteItem(fakeEvent, '-idB-');
    expect(onCompleteItem).toBeCalledWith('-idB-');
  });

  it('Should show todo items', () => {
    const initialState = {
      items: [{ title: 'Test mapStateToProps', id: '-idMSTP-', isCompleted: false }]
    };
    expect(mapStateToProps(initialState)).toEqual({
      items: [{ title: 'Test mapStateToProps', id: '-idMSTP-', isCompleted: false }]
    });
  });

  it('Should dispatch onAddItem successfully', () => {
    const dispatch = jest.fn();
    mapDispatchToProps(dispatch).onAddItem('dispatch');
    expect(dispatch).toBeCalledWith({ item: 'dispatch', type: 'ADD_ITEM' });
  });

  it('Should dispatch onCompleteItem successfully', () => {
    const dispatch = jest.fn();
    mapDispatchToProps(dispatch).onCompleteItem('idA');
    expect(dispatch).toBeCalledWith({ id: 'idA', type: 'COMPLETE_ITEM' });
  });

  it('Should dispatch onDeleteItem successfully', () => {
    const dispatch = jest.fn();
    mapDispatchToProps(dispatch).onDeleteItem('idB');
    expect(dispatch).toBeCalledWith({ id: 'idB', type: 'DELETE_ITEM' });
  });
});
