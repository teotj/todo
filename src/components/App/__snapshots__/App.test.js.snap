// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App> Should render 1`] = `
<div
  className="root"
>
  <Component
    className="header"
  >
    Todo
  </Component>
  <ErrorBoundary>
    <Blueprint3.FormGroup
      className="formGroup"
    >
      <Blueprint3.InputGroup
        className="todoItem"
        large={true}
        onChange={[Function]}
        placeholder="Add todo here..."
        rightElement={null}
        value=""
      />
    </Blueprint3.FormGroup>
    <Blueprint3.Button
      className="add"
      disabled={true}
      icon="add"
      intent="success"
      onClick={[Function]}
      text="Add todo"
    />
  </ErrorBoundary>
  <div
    className="todoTable"
  >
    <ErrorBoundary>
      <Blueprint3.Tabs
        animate={true}
        className="tabs"
        id="todoTabs"
        large={false}
        renderActiveTabPanelOnly={true}
        vertical={false}
      >
        <Blueprint3.Tab
          className="tab"
          disabled={false}
          id="all"
          key="all"
          panel={
            <TodoTable
              items={
                Array [
                  Object {
                    "id": "-idA-",
                    "isCompleted": true,
                    "title": "TodoA",
                  },
                  Object {
                    "id": "-idB-",
                    "isCompleted": false,
                    "title": "TodoB",
                  },
                ]
              }
              onComplete={[Function]}
              onDelete={[Function]}
            />
          }
          title="All"
        />
        <Blueprint3.Tab
          className="tab"
          disabled={false}
          id="completed"
          key="completed"
          panel={
            <TodoTable
              items={
                Array [
                  Object {
                    "id": "-idA-",
                    "isCompleted": true,
                    "title": "TodoA",
                  },
                ]
              }
              onComplete={[Function]}
              onDelete={[Function]}
            />
          }
          title="Completed"
        />
        <Blueprint3.Tab
          className="tab"
          disabled={false}
          id="active"
          key="active"
          panel={
            <TodoTable
              items={
                Array [
                  Object {
                    "id": "-idB-",
                    "isCompleted": false,
                    "title": "TodoB",
                  },
                ]
              }
              onComplete={[Function]}
              onDelete={[Function]}
            />
          }
          title="Active"
        />
      </Blueprint3.Tabs>
    </ErrorBoundary>
  </div>
</div>
`;
